File: macos/Sources/ElevateCore/Managed/ManagedConfiguration.swift
import Foundation

/// Settings pushed by an organization's MDM, merged from whichever `ManagedConfigurationSource`
/// the app is configured with. Every field is optional or defaulted so an app with no managed
/// configuration behaves exactly as it would without this type existing at all.
public struct ManagedConfiguration: Hashable, Sendable {
    public var clientId: String?
    public var disableUpdateCheck: Bool
    public var allowedSignInMethods: Set<SignInMethodKind>?
    public var allowedTenants: [String]?
    public var pinnedTenants: [String]
    /// Raw JSON text for the managed profiles document; parsed elsewhere.
    public var managedProfilesDocument: String?
    public var managedProfilesUrl: URL?
    /// The organization's display name, 1-32 characters. Gates the three keys below it.
    public var organizationName: String?
    public var organizationTitleStyle: OrganizationTitleStyle?
    public var organizationSupportUrl: URL?
    public var organizationSupportEmail: String?
    /// The keys that carried a valid, in-effect value, in `ManagedKey.allCases` order.
    public var keysInEffect: [ManagedKey]
    /// Human-readable notes about values that were present but rejected.
    public var warnings: [String]
    /// The source's `origin`, set only when at least one key is in effect.
    public var origin: String?

    public init() {
        clientId = nil
        disableUpdateCheck = false
        allowedSignInMethods = nil
        allowedTenants = nil
        pinnedTenants = []
        managedProfilesDocument = nil
        managedProfilesUrl = nil
        organizationName = nil
        organizationTitleStyle = nil
        organizationSupportUrl = nil
        organizationSupportEmail = nil
        keysInEffect = []
        warnings = []
        origin = nil
    }

    /// The configuration with nothing managed — equivalent to `ManagedConfiguration()`.
    public static let none = ManagedConfiguration()

    public var isEmpty: Bool { keysInEffect.isEmpty }

    /// Reads and validates every managed key from `source`, collecting warnings for values that
    /// were present but rejected. Keys with no value, or with a value that fails validation, are
    /// simply absent from the result — they do not affect `keysInEffect` or `origin`.
    public static func load(from source: any ManagedConfigurationSource) -> ManagedConfiguration {
        var config = ManagedConfiguration()

        if let raw = source.string(.clientId) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if let uuid = UUID(uuidString: trimmed), uuid != UUID(uuidString: "00000000-0000-0000-0000-000000000000") {
                config.clientId = trimmed.lowercased()
                config.keysInEffect.append(.clientId)
            } else {
                config.warnings.append("ClientId: '\(raw)' is not a valid GUID")
            }
        }

        if let flag = source.bool(.disableUpdateCheck) {
            config.disableUpdateCheck = flag
            config.keysInEffect.append(.disableUpdateCheck)
        }

        if let names = source.list(.allowedSignInMethods), !names.isEmpty {
            var kinds: Set<SignInMethodKind> = []
            for name in names {
                if let kind = SignInMethodKind.allCases.first(where: { $0.rawValue.lowercased() == name.lowercased() }) {
                    kinds.insert(kind)
                } else {
                    config.warnings.append("AllowedSignInMethods: unknown method '\(name)' ignored")
                }
            }
            if !kinds.isEmpty {
                config.allowedSignInMethods = kinds
                config.keysInEffect.append(.allowedSignInMethods)
            }
        }

        if let tenants = source.list(.allowedTenants) {
            let cleaned = Self.cleanedTenantList(tenants)
            if !cleaned.isEmpty {
                config.allowedTenants = cleaned
                config.keysInEffect.append(.allowedTenants)
            }
        }

        if let tenants = source.list(.pinnedTenants) {
            let cleaned = Self.cleanedTenantList(tenants)
            if !cleaned.isEmpty {
                config.pinnedTenants = cleaned
                config.keysInEffect.append(.pinnedTenants)
            }
        }

        if let raw = source.string(.managedProfiles) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if !trimmed.isEmpty {
                config.managedProfilesDocument = raw
                config.keysInEffect.append(.managedProfiles)
            }
        }

        if let raw = source.string(.managedProfilesUrl) {
            if let url = URL(string: raw), url.scheme == "https" {
                config.managedProfilesUrl = url
                config.keysInEffect.append(.managedProfilesUrl)
            } else {
                config.warnings.append("ManagedProfilesUrl: only https URLs are accepted")
            }
        }

        if let raw = source.string(.organizationName) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty {
                config.warnings.append("OrganizationName: a blank name is ignored")
            } else if trimmed.count > 32 {
                config.warnings.append("OrganizationName: '\(trimmed)' is \(trimmed.count) characters; the maximum is 32")
            } else {
                config.organizationName = trimmed
                config.keysInEffect.append(.organizationName)
            }
        }

        // The three keys below decorate the name; without one they have nothing to attach to.
        let hasName = config.organizationName != nil

        if let raw = source.string(.organizationTitleStyle) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if !hasName {
                config.warnings.append("OrganizationTitleStyle: ignored because OrganizationName is not set")
            } else if let style = OrganizationTitleStyle.allCases.first(where: { $0.rawValue.lowercased() == trimmed.lowercased() }) {
                config.organizationTitleStyle = style
                config.keysInEffect.append(.organizationTitleStyle)
            } else {
                config.warnings.append("OrganizationTitleStyle: '\(raw)' is not one of by, managedBy, none")
            }
        }

        if let raw = source.string(.organizationSupportUrl) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if !hasName {
                config.warnings.append("OrganizationSupportUrl: ignored because OrganizationName is not set")
            } else if let url = URL(string: trimmed), url.scheme == "https" {
                config.organizationSupportUrl = url
                config.keysInEffect.append(.organizationSupportUrl)
            } else {
                config.warnings.append("OrganizationSupportUrl: only https URLs are accepted")
            }
        }

        if let raw = source.string(.organizationSupportEmail) {
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if !hasName {
                config.warnings.append("OrganizationSupportEmail: ignored because OrganizationName is not set")
            } else if trimmed.count >= 3, trimmed.contains("@"), !trimmed.contains(where: \.isWhitespace) {
                config.organizationSupportEmail = trimmed
                config.keysInEffect.append(.organizationSupportEmail)
            } else {
                config.warnings.append("OrganizationSupportEmail: '\(raw)' is not an email address")
            }
        }

        if !config.keysInEffect.isEmpty {
            config.origin = source.origin
        }

        return config
    }

    /// Trims, lower-cases, drops blanks, and de-duplicates while preserving order.
    private static func cleanedTenantList(_ raw: [String]) -> [String] {
        var seen = Set<String>()
        var result: [String] = []
        for entry in raw {
            let cleaned = entry.trimmingCharacters(in: .whitespaces).lowercased()
            guard !cleaned.isEmpty, !seen.contains(cleaned) else { continue }
            seen.insert(cleaned)
            result.append(cleaned)
        }
        return result
    }
}

File: windows/src/Elevate.Core/Managed/ManagedConfiguration.cs
using Elevate.Core.Models;

namespace Elevate.Core.Managed;

/// <summary>
/// Settings pushed by an organization's policy, merged from whichever <see cref="IManagedConfigurationSource"/>
/// the app is configured with. Every property is optional or defaulted so an app with no managed
/// configuration behaves exactly as it would without this type existing at all. Port of the Swift
/// <c>ManagedConfiguration</c> struct.
/// </summary>
public sealed record ManagedConfiguration
{
    public string? ClientId { get; init; }

    public bool DisableUpdateCheck { get; init; }

    public IReadOnlySet<SignInMethodKind>? AllowedSignInMethods { get; init; }

    public IReadOnlyList<string>? AllowedTenants { get; init; }

    public IReadOnlyList<string> PinnedTenants { get; init; } = [];

    /// <summary>Raw JSON text for the managed profiles document; parsed elsewhere.</summary>
    public string? ManagedProfilesDocument { get; init; }

    public Uri? ManagedProfilesUrl { get; init; }

    /// <summary>The keys that carried a valid, in-effect value, in <see cref="ManagedKeys.All"/> order.</summary>
    public IReadOnlyList<ManagedKey> KeysInEffect { get; init; } = [];

    /// <summary>Human-readable notes about values that were present but rejected.</summary>
    public IReadOnlyList<string> Warnings { get; init; } = [];

    /// <summary>The source's <see cref="IManagedConfigurationSource.Origin"/>, set only when at least one key is in effect.</summary>
    public string? Origin { get; init; }

    /// <summary>The configuration with nothing managed — equivalent to <c>new ManagedConfiguration()</c>.</summary>
    public static ManagedConfiguration None { get; } = new();

    public bool IsEmpty => KeysInEffect.Count == 0;

    /// <summary>
    /// Reads and validates every managed key from <paramref name="source"/>, collecting warnings
    /// for values that were present but rejected. Keys with no value, or with a value that fails
    /// validation, are simply absent from the result — they do not affect <see cref="KeysInEffect"/>
    /// or <see cref="Origin"/>.
    /// </summary>
    public static ManagedConfiguration Load(IManagedConfigurationSource source)
    {
        ArgumentNullException.ThrowIfNull(source);

        string? clientId = null;
        var disableUpdateCheck = false;
        HashSet<SignInMethodKind>? allowedSignInMethods = null;
        IReadOnlyList<string>? allowedTenants = null;
        IReadOnlyList<string> pinnedTenants = [];
        string? managedProfilesDocument = null;
        Uri? managedProfilesUrl = null;
        var keysInEffect = new List<ManagedKey>();
        var warnings = new List<string>();

        var rawClientId = source.String(ManagedKey.ClientId);
        if (rawClientId is not null)
        {
            var trimmed = rawClientId.Trim();
            if (Guid.TryParseExact(trimmed, "D", out var guid) && guid != Guid.Empty)
            {
                clientId = guid.ToString().ToLowerInvariant();
                keysInEffect.Add(ManagedKey.ClientId);
            }
            else
            {
                warnings.Add($"ClientId: '{rawClientId}' is not a valid GUID");
            }
        }

        var flag = source.Bool(ManagedKey.DisableUpdateCheck);
        if (flag is not null)
        {
            disableUpdateCheck = flag.Value;
            keysInEffect.Add(ManagedKey.DisableUpdateCheck);
        }

        var methodNames = source.List(ManagedKey.AllowedSignInMethods);
        if (methodNames is { Count: > 0 })
        {
            var kinds = new HashSet<SignInMethodKind>();
            foreach (var name in methodNames)
            {
                if (Enum.TryParse<SignInMethodKind>(name, ignoreCase: true, out var kind) && Enum.IsDefined(kind))
                {
                    kinds.Add(kind);
                }
                else
                {
                    warnings.Add($"AllowedSignInMethods: unknown method '{name}' ignored");
                }
            }

            if (kinds.Count > 0)
            {
                allowedSignInMethods = kinds;
                keysInEffect.Add(ManagedKey.AllowedSignInMethods);
            }
        }

        var allowedTenantsRaw = source.List(ManagedKey.AllowedTenants);
        if (allowedTenantsRaw is not null)
        {
            var cleaned = CleanTenantList(allowedTenantsRaw);
            if (cleaned.Count > 0)
            {
                allowedTenants = cleaned;
                keysInEffect.Add(ManagedKey.AllowedTenants);
            }
        }

        var pinnedTenantsRaw = source.List(ManagedKey.PinnedTenants);
        if (pinnedTenantsRaw is not null)
        {
            var cleaned = CleanTenantList(pinnedTenantsRaw);
            if (cleaned.Count > 0)
            {
                pinnedTenants = cleaned;
                keysInEffect.Add(ManagedKey.PinnedTenants);
            }
        }

        var rawProfiles = source.String(ManagedKey.ManagedProfiles);
        if (rawProfiles is not null)
        {
            var trimmed = rawProfiles.Trim();
            if (trimmed.Length > 0)
            {
                managedProfilesDocument = rawProfiles;
                keysInEffect.Add(ManagedKey.ManagedProfiles);
            }
        }

        var rawUrl = source.String(ManagedKey.ManagedProfilesUrl);
        if (rawUrl is not null)
        {
            if (Uri.TryCreate(rawUrl, UriKind.Absolute, out var uri) && uri.Scheme == Uri.UriSchemeHttps)
            {
                managedProfilesUrl = uri;
                keysInEffect.Add(ManagedKey.ManagedProfilesUrl);
            }
            else
            {
                warnings.Add("ManagedProfilesUrl: only https URLs are accepted");
            }
        }

        return new ManagedConfiguration
        {
            ClientId = clientId,
            DisableUpdateCheck = disableUpdateCheck,
            AllowedSignInMethods = allowedSignInMethods,
            AllowedTenants = allowedTenants,
            PinnedTenants = pinnedTenants,
            ManagedProfilesDocument = managedProfilesDocument,
            ManagedProfilesUrl = managedProfilesUrl,
            KeysInEffect = keysInEffect,
            Warnings = warnings,
            Origin = keysInEffect.Count > 0 ? source.Origin : null,
        };
    }

    /// <summary>Trims, lower-cases, drops blanks, and de-duplicates while preserving order.</summary>
    private static List<string> CleanTenantList(IReadOnlyList<string> raw)
    {
        var seen = new HashSet<string>();
        var result = new List<string>();
        foreach (var entry in raw)
        {
            var cleaned = entry.Trim().ToLowerInvariant();
            if (cleaned.Length == 0 || !seen.Add(cleaned))
            {
                continue;
            }

            result.Add(cleaned);
        }

        return result;
    }
}
